refactor(colors): Rework `colors` structure

magistau opened the pull request:
- Add support for ANSI 4-bit colors
- Fix `-dec` computation

magistau committed:
refactor(colors): Rework `colors` structure

- Add support for ANSI 4-bit colors
- Fix `-dec` computation

SenchoPens commented:
Good evening, thank you for your PR.

I have a few questions:
1. What was the problem in `-dec` computation?
2. What use cases are there for ansi colors, other than wezterm hm option?
3. Can you explain the overall motivation behind it? It seems more complex than the previous version, though I am biased :)
4. Is it backward compatible?

magistau commented:
> 1. What was the problem in `-dec` computation?

A `-dec` value of `1.0` corresponds to a `-rgb` value of `255`; prior to this change, it would instead correspond to `256`. Basically, the fix is replacing `/ 256.0` by `/ 255.0`,

> 2. What use cases are there for ansi colors, other than wezterm hm option?

WezTerm was my motivation, since I use it; another example is Alacritty which wouldn't even need `toList`:
```nix
programs.alacritty.settings.colors = let inherit (builtins) removeAttrs; in {
  normal = removeAttrs scheme.withHashtag.ansi.dark ["toList"];
  bright = removeAttrs scheme.withHashtag.ansi.bright ["toList"];
};
```
(to be honest, I only found out how that is configured while writing this comment; with that in mind, renaming `ansi.dark` to `ansi.normal` might make sense)

Actually, anywhere you see a terminal you would want to use `scheme.ansi`, would that be a terminal emulator, an editor with an integrated terminal (IBNLT VSCode) or, well, `console.colors`. Note that the old documentation about `scheme.toList` being good for `console.colors` is not quite right: the first eight base16 colors are shades of gray and as such are not a good fit for 3-bit ANSI color palette; the latter eight colors are not a good fit either, since the bright 4-bit ANSI colors do not have orange or brown, but include "bright black" and "bright white" colors.
See the difference for yourself:
base16 palette | 4-bit ANSI palette
-------------- | ------------------
![base 16 palette](https://github.com/SenchoPens/base16.nix/assets/43097806/a5fff6ff-9969-48ae-8f00-51de172dde3d) | ![4-bit ANSI palette](https://github.com/SenchoPens/base16.nix/assets/43097806/192bf8d6-11aa-46d4-812f-dbbff9de091d)

You might've noticed that I updated the doc comment for `colors` function: it has a link to the [corresponding documentation of base24](https://github.com/tinted-theming/base24/blob/master/styling.md#specific-colours-and-their-usages).

> 3. Can you explain the overall motivation behind it? It seems more complex than the previous version, though I am biased :)

I made a `_color` function for creating a color structure from its hex value, which allows for bringing all the color converting logic in one place and makes the structure more evident; for backwards compatibility everything is eventually `toString`ed, with a new `original` attribute being an attrset of non-flattened colors.
I have also formatted the file with `nixfmt-rfc-style` since I wanted to keep a uniform code style in the file, and I'm pretty used to that one; if you have an actual code style, I'm happy to use that. Some of that complexity might arise from the code style, e.g. lists with more than one element are forced to be multiline.
I have also rewritten the flattening logic: instead of several function for different `-suffix`es there's a single function (`based`) which uses a list of paths and creates a new attrset with every `path-to-attr` referring to `path.to.attr` of the source attrset (`base-colors`).

> 4. Is it backward compatible?

There are two theoretically breaking changes:
- `-dec` values are different from the old ones by a factor of $\frac{256}{255}$
- `withHashtag` only affects attributes of form `baseXX-hex`; previously, all attributes (including e.g. `-dec-X` and `-hex-g`) would be prefixed with a hashtag which was quite surprising for me; the only actual problem might be with the `-hex-bgr` attribute, which is not `#`-prefixed anymore; I don't know of any software using bgr format and I did not find any reference to bgr in stylix (which is the only project depending on base16.nix I'm aware of), so I have no idea whether `#BBGGRR` format makes sense or not.

SenchoPens commented on lib/colors.nix:
```diff
@@ -1,65 +1,76 @@
+      (map (lib.flip getAttr digits'))
+      (lib.foldr (cur: acc: acc * base + cur) 0)
+    ];
+  toIntBase16 = parseIntRadix 16;
```
Does the `toIntBase16` function do the same as `primaryHex2Dec`? While the old name may not be the best, the old implementation is a lot simpler. Are there any benefits to your implementation to justify the complexity?

magistau replied:
`primaryHex2Dec` looks too specific; this implementation makes the intent more clear.

SenchoPens replied:
What do you mean by clear and specific? Does it handle some existing usecase not handled by the old implementation? The code does not have any tests so I still prefer the old one as the only logic it has is 2 attrset lookups.

magistau replied:
I could add some tests if needed

SenchoPens replied:
Overall, the library would benefit from tests, but this is one of the two cons. The second con is that I do not see how a composition of ~10 functions state the intent clearer than 2 attrset lookups.

SenchoPens commented on lib/colors.nix:
```diff
@@ -69,57 +80,204 @@ let
+            toList = ansi-list self.bright;
+          };
+          # suitable for console.colors and similar
+          toList = self.dark.toList ++ self.bright.toList;
```
This changes the order and the new one is the correct one, right?

magistau replied:
This is a new attribute, it does not change any order as it did not exist prior to this change

SenchoPens commented on lib/colors.nix:
```diff
@@ -69,57 +80,204 @@ let
+        }
+        // builtins.listToAttrs (
+          lib.zipListsWith lib.nameValuePair
+            (map (x: "bright-${x}") [
```
just prepend bright- to each color

magistau replied:
This is exactly what happens, just without boilerplate

SenchoPens replied:
Sadly, the code does not have any tests so more boilerplate is better than more logic

SenchoPens commented:
The changes make sense, I've added some small comments. Maybe you can replace the point-free style with plain lambdas? I'm not that familiar with the nixpkgs library and sometimes the overhaul of compositions looks unnecessary complex

magistau commented:
> The changes make sense, I've added some small comments. Maybe you can replace the point-free style with plain lambdas? I'm not that familiar with the nixpkgs library and sometimes the overhaul of compositions looks unnecessary complex

I can agree that sometimes the point-freeness is too extreme, will change.

magistau committed:
refactor: (hopefully) make the code more intuitive

SenchoPens commented:
Hi, I've left a couple of comments

SenchoPens commented on lib/colors.nix:
```diff
@@ -69,57 +80,171 @@ let
+      total = palette: based palette // extra palette;
+      total' =
+        palette:
+        mapRecursiveCond (x: lib.strings.isConvertibleWithToString x -> isList x) toString (total palette)
```
Can you please explain what does this line do?

magistau replied:
First of all it takes `total palette`; after that it transforms recursively:
1. If the value under question is not `toString`-able or is a list, recurse with `mapAttrs` or `map`
2. Otherwise use `toString` to turn it into a string

Effectively this transforms all color nodes into strings (they have a `__toString` attr so they are `toString`-able) while keeping the attrset (e.g. `mnemonic`, `ansi`, `ansi.bright`) and list (e.g. `ansi.dark.toList`) structure

SenchoPens replied:
This looks like a workaround. I think you can remove both `__toString` and instead use a common attribute like `.asString`. Then it will look like `mapRecursiveCond (x: x ? asString) (x: x.asString) (total palette)`.

As far as I understood, you use the implicit conversion to string in only in one other place - in the code for the `_color` function. There you can do without this conversion: instead of `self.hex` just use the `hex` parameter. This will also make for a simpler control flow.

magistau replied:
My previous comment was incomplete: this transformation also affects numbers such as `base01-dec-r` and `base0E-rgb-b`, so `lib.strings.isConvertibleWithToString x -> isList x` is not really simplifiable (an alternative is `isFloat x || isInt x || (isAttrs x && x ? hex)`)

SenchoPens replied:
Then maybe we can convert them to string in the _color function itself?

magistau replied:
They were specifically implemented in the way they currently are to make scripting easier; making all `_color` attributes strings would be against that intention

SenchoPens replied:
I am a bit confused. By scripting do you mean user-level scripting? Because you still convert them to string, so for the user it does not matter.

magistau replied:
A user might want to do calculations with colors (including but not limited to color space conversion) and parsing stringified numbers is suboptimal. On a second thought, the color constructor should perhaps be made public for this reason.

SenchoPens replied:
Ok, but in your PR they still get converted to string, don't they? We can explicitly convert them to string as I proposed above, and provide additional attributes with the integer values. @thecaralice

magistau replied:
Not sure what you mean, they are converted to strings to match base16/24 spec, and they have the additional attributes in the `original` attrset.

SenchoPens replied:
Oh, I got it. After another thought, the actual problem is the following. You open too much of the internals of the library (increasing the future maintenance burden) to save a little time for 1% of "power users".

But even further, even this potential time saving is under question. Right now, the features you've added are undocumented, so nobody will know about them, thus they are code bloat. Even if we add documentation, it would look like "remember that dec-r attribute? We called `lib.strings.toInt` for you and you can access this value via `original.base00.dec-r`", being another sort of bloat.

@thecaralice

SenchoPens replied:
In any case, as I've stated in the README, the primary purpose of base16.nix is to be an interface to the base16 standard. Any additional features I want to keep at minimum.

SenchoPens replied:
BTW, can you please document the featuers you've added in the DOCUMENTATION.md?

SenchoPens reviewed (changes requested):
Left one more comment

SenchoPens reviewed (changes requested):
Hi, added a couple of comments

SenchoPens reviewed (commented):
Looks great, left one comment.

PS sorry for the wait, I was on a trip.
